### PR TITLE
feat(isp): support switch for launch ag and forward overlap per module

### PR DESCRIPTION
**Pull request by huangting4201**
## Motivation

Optimize isp wp overlap to adapt to moe and different seqlen and parallel cases.
1. support user switch for setting wp forward allgather launch module.
2. support user switch for setting wp forward overlap granularity, forward overlap per 'layer' or 'module'. When overlap by module, please set env `export CUDA_DEVICE_MAX_CONNECTIONS=1`.

```python
"""
weight parallel (dict):
    1. size: int, the size of weight parallel.
    2. overlap: bool, enable/disable all_gather/reduce_scatter communication overlap, defaults to False.
    3. launch_allgather_before: str, before which module to launch the all gather communication to
        prefetch next layer's weight, should be in ['wqkv', 'attn', 'wo', 'w1'], defaults to 'wo'.
        Must be used with forward_overlap_per 'layer'.
    4. forward_overlap_per: str, all gather prefetch granularity, per 'module' or per 'layer', defaults to 'layer'.
"""
parallel = dict(
    zero1=dict(size=-1),
    tensor=dict(size=2, mode="isp"),
    pipeline=dict(size=1, interleaved_overlap=True),
    weight=dict(size=4, overlap=True, launch_allgather_before="wo", forward_overlap_per="layer"),
)
```

## Modification

`internlm/core/parallel/comm/isp.py: ISPCommunicator`.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/7B_MoE4_sft.py [hunk 1]
```diff
@@ -183,6 +183,10 @@
 weight parallel (dict):
     1. size: int, the size of weight parallel.
     2. overlap: bool, enable/disable all_gather/reduce_scatter communication overlap, defaults to False.
+    3. launch_allgather_before: str, before which module to launch the all gather communication to
+        prefetch next layer's weight, should be in ['wqkv', 'attn', 'wo', 'w1'], defaults to 'wo'.
```
Comment (yingtongxiong) [inline configs/7B_MoE4_sft.py:187]: 这里为啥没有w2，w3呢？担心不够overlap住？

File: internlm/core/parallel/comm/isp.py [hunk 2]
```diff
@@ -369,7 +404,7 @@ def get_model(obj: nn.Module) -> nn.Module:
                     self._overlap_states[cid].index_to_isp_modules[idx] = []
                     self._overlap_states[cid].index_to_block[idx] = block
                     for name, child in block.named_modules():
-                        if name.split(".")[-1] in ["out_proj", "wo"]:
+                        if is_allgather_launch_module(name, child):
                             self._overlap_states[cid].isp_outs.append(child)
```
Comment (yingtongxiong) [inline internlm/core/parallel/comm/isp.py]: isp_outs这个名字要改一下吗？我理解这里放的都是self._launch_before_mudole指定的module？

File: configs/7B_MoE4_sft.py (hunk 1 above)
Comment (huangting4201) [reply configs/7B_MoE4_sft.py:187]: 因为觉得没有必要这么细，没有必要给用户这么多选择， ['wqkv', 'attn', 'wo', 'w1']应该已经够用

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py]: 改一下吧，好一点

File: configs/7B_MoE4_sft.py (hunk 1 above)
Comment (yingtongxiong) [reply configs/7B_MoE4_sft.py:187]: 好的

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (yingtongxiong) [reply internlm/core/parallel/comm/isp.py]: 好的

Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py]: 已更新：https://github.com/InternLM/InternEvo/pull/381/commits/e53c5f078ea423e4284feb4e83800e8fef1e2927

**Review by yingtongxiong — APPROVED** [review]

**Review by sunpengsdu — APPROVED** [review]

**Pull request merged**